File: 2015/day20.py
from typing import Callable
from os.path import isfile, join as path_join

file_name = path_join('input', 'day20.txt')


def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]


def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)


if not isfile(file_name):
	from aoc import get_input_file

	get_input_file()

from math import sqrt, floor

fact = [1, 2, 6, 24, 120, 720, 5040, 40320, 362880, 3628800, 39916800, 479001600, 6227020800, 87178291200]

number = int(open(file_name).read().strip())


def divisor_sum_both(n):  # incorrect for n=1
	max_n = floor(sqrt(n))
	s1 = n + 1
	s2 = n + 1
	for i in range(2, max_n + 1):
		if n % i == 0:
			s1 += i
			if i * 50 >= n:
				s2 += i
			d = n // i
			if d != i:
				s1 += d
				if d * 50 >= n:
					s2 += d

	return s1, s2


def run_both_parts():
	# 4 => 1*10 + 2*10 + 4*10 = (1+2+4)*10
	# divisors of 4 are 1,2,4
	# so 4 => div-sum * 10
	min_div_sum1 = number // 10
	min_div_sum2 = number // 11

	found1 = False
	found2 = False

	answer1 = 0
	answer2 = 0

	# given that div_sum(n!) has the highest divisor sum from 1->n!
	# we can use use div_sum(n!) to find n1 and n2 where ds(n1!) <= ds(t) <= ds(n2!)

	min_target = min(min_div_sum1, min_div_sum2)
	start_i = 1
	for i in range(len(fact) - 1):
		if divisor_sum_both(fact[i])[1] <= min_target < divisor_sum_both(fact[i + 1])[0]:
			start_i = fact[i]
			break

	i = start_i
	while True:
		s = divisor_sum_both(i)
		# print(s)
		if not found1 and s[0] >= min_div_sum1:
			answer1 = i
			found1 = True

		if not found2 and s[1] >= min_div_sum2:
			answer2 = i
			found2 = True

		if found1 and found2:
			break

		i += 1

	return answer1, answer2


solution = run_both_parts()


def part1():
	return solution[0]


def part2():
	return solution[1]


print(part1())
print(part2())

File: cpp/2015/day20.cpp
#include "../aocHelper.h"

constexpr long long factorial(long long n)
{
	if (n <= 1)
	{
		return 1;
	}
	return n * factorial(n - 1);
}

class Day20 : public BaseDay
{
public:
	Day20() : BaseDay("20") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		static constexpr int FACTS_COUNT = 14;

		static constexpr long long facts[FACTS_COUNT] = { factorial(1), factorial(2), factorial(3), factorial(4), factorial(5), factorial(6), factorial(7), factorial(8), factorial(9), factorial(10), factorial(11), factorial(12), factorial(12), factorial(14) };

		static_assert(facts[3] == 24);

		// incorrect for n=1
		auto divisor_sum_both = [](long long n)
		{
			long long max_n = floor(sqrt(n));

			auto result = make_pair(n + 1, n + 1);

			for (long long i = 2; i <= max_n; i++)
			{
				if (n % i == 0)
				{
					result.first += i;
					if (i * 50 >= n)
					{
						result.second += i;
					}
					long long d = n / i;
					if (d != i)
					{
						result.first += d;
						if (d * 50 >= n)
						{
							result.second += d;
						}
					}
				}
			}

			return result;
		};

		auto run_both_parts = [&divisor_sum_both](long long number)
		{
			// 4 = > 1 * 10 + 2 * 10 + 4 * 10 = (1 + 2 + 4) * 10
			// divisors of 4 are 1, 2, 4
			// so 4 = > div - sum * 10
			auto min_div_sum = make_pair(number / 10, number / 11);

			auto found = make_pair(false, false);

			auto result = make_pair(0, 0);

			// given that div_sum(n!) has the highest divisor sum from 1->n!
			// we can use use div_sum(n!) to find n1 and n2 where ds(n1!) <= ds(t) <= ds(n2!)

			long long min_target = min(min_div_sum.first, min_div_sum.second);
			long long start_i = 1;

			for (int i = 0; i < FACTS_COUNT - 1; i++)
			{
				if (divisor_sum_both(facts[i]).second <= min_target && min_target < divisor_sum_both(facts[i + 1]).first)
				{
					start_i = facts[i];
					break;
				}
			}

			long long i = start_i;

			while (true)
			{
				auto s = divisor_sum_both(i);

				if (!found.first && s.first >= min_div_sum.first)
				{
					result.first = i;
					found.first = true;
				}

				if (!found.second && s.second >= min_div_sum.second)
				{
					result.second = i;
					found.second = true;
				}

				if (found.first && found.second)
				{
					break;
				}

				i++;
			}

			return result;
		};

		// parse input
		int number = numericParse<long long>(input);

		auto solution = run_both_parts(number);

		// part 1
		part1 = solution.first;

		// part 2
		part2 = solution.second;

		return { part1, part2 };
	}
};
